### PR TITLE
Skip python interpreter discovery for 'forced local' module execution (#74824)

### PR DESCRIPTION
**Pull request by bcoca**
Backport of #74824 for Ansible 2.11.

  mostly for use with network_os
  use 'remote is local' property as indicator
  ensure task_vars are as expected in test

(cherry picked from commit 8d41b97329cae281ce194dbb8cb3ce35fdce23ec)

##### ISSUE TYPE

- Bugfix Pull Request

##### COMPONENT NAME

core

### CODE REVIEW

**Comment by sivel** [comment]
It looks like this PR has uncovered an encoding issue in the `find` module.  Since `vyos` presents a normal linux environment, I'm guessing that previously interpreter discovery was selecting a Python binary that didn't cause the issue.  It looks like based on the last nightly CI run, it's using `/usr/bin/python`:

```
01:28 TASK [incidental_vyos_config : setup] ******************************************
01:32 changed: [vyos-1-1-8] => {"ansible_facts": {"discovered_interpreter_python": "/usr/bin/python"}, "changed": true, "commands": ["set system host-name vyos-1-1-8"], "filtered": []}
```

`/usr/bin/python -V` on the vyos host reports `Python 2.6.6`

But without discovery, we're using `/usr/bin/python2.7` which is the python used to spawn `ansible-playbook`.

```
02:18 Traceback (most recent call last):
02:18   File "/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py", line 126, in <module>
02:18     _ansiballz_main()
02:18   File "/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py", line 118, in _ansiballz_main
02:18     invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
02:18   File "/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py", line 67, in invoke_module
02:18     run_name='__main__', alter_sys=True)
02:18   File "/usr/lib/python2.7/runpy.py", line 188, in run_module
02:18     fname, loader, pkg_name)
02:18   File "/usr/lib/python2.7/runpy.py", line 82, in _run_module_code
02:18     mod_name, mod_fname, mod_loader, pkg_name)
02:18   File "/usr/lib/python2.7/runpy.py", line 72, in _run_code
02:18     exec code in run_globals
02:18   File "/tmp/ansible_find_payload_EVf5zz/ansible_find_payload.zip/ansible/modules/find.py", line 505, in <module>
02:18   File "/tmp/ansible_find_payload_EVf5zz/ansible_find_payload.zip/ansible/modules/find.py", line 496, in main
02:18 UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 83: ordinal not in range(128)
02:18 fatal: [vyos-1-1-8]: FAILED! => {
02:18     "changed": false, 
02:18     "module_stderr": "Traceback (most recent call last):\n  File \"/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py\", line 126, in <module>\n    _ansiballz_main()\n  File \"/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py\", line 118, in _ansiballz_main\n    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)\n  File \"/root/.ansible/tmp/ansible-local-4645u8mIU/ansible-tmp-1628173102.8-728-174442077434566/AnsiballZ_find.py\", line 67, in invoke_module\n    run_name='__main__', alter_sys=True)\n  File \"/usr/lib/python2.7/runpy.py\", line 188, in run_module\n    fname, loader, pkg_name)\n  File \"/usr/lib/python2.7/runpy.py\", line 82, in _run_module_code\n    mod_name, mod_fname, mod_loader, pkg_name)\n  File \"/usr/lib/python2.7/runpy.py\", line 72, in _run_code\n    exec code in run_globals\n  File \"/tmp/ansible_find_payload_EVf5zz/ansible_find_payload.zip/ansible/modules/find.py\", line 505, in <module>\n  File \"/tmp/ansible_find_payload_EVf5zz/ansible_find_payload.zip/ansible/modules/find.py\", line 496, in main\nUnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 83: ordinal not in range(128)\n", 
02:18     "module_stdout": "", 
02:18     "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error", 
02:18     "rc": 1
02:18 }
```

I don't know if we need to set an `ansible_python_interpreter` here in the tests for `vyos` on stable-2.11, or whether we should resolve the `find` issue first.

**Comment by bcoca** [comment]
both? I'll look into the find module

File: lib/ansible/modules/find.py [hunk 1]
```diff
@@ -460,21 +460,22 @@ def main():
                         continue
 
                     r = {'path': fsname}
-                    if params['file_type'] == 'any':
+                    file_type = to_text(params['file_type'], errors='surrogate_or_strict')
```
Comment (nitzmahone) [inline lib/ansible/modules/find.py]: unrelated changes in this file

**Review by nitzmahone — CHANGES_REQUESTED** [review]
looks like some extra cruft from other changes snuck in here

File: lib/ansible/modules/find.py (hunk 1 above)
Comment (sivel) [reply lib/ansible/modules/find.py]: It's partially related to my other comment you can see here https://github.com/ansible/ansible/pull/75402#issuecomment-893543887

Comment (bcoca) [reply lib/ansible/modules/find.py]: just testing since PR is 'reproducer', but it seems to no t be the issue

**Comment by bcoca** [comment]
Once linked PR above is merged, this should be able to pass vyos test

**Review by nitzmahone — APPROVED** [review]
